File: common/src/main/java/com/hibiscusmc/hmcleaves/world/LeavesWorld.java
package com.hibiscusmc.hmcleaves.world;

import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.UnmodifiableView;

import java.util.Collections;
import java.util.Map;
import java.util.UUID;
import java.util.function.Consumer;
import java.util.function.Function;

public final class LeavesWorld {

    private final UUID worldId;
    private final Map<ChunkPosition, LeavesChunk> chunks;

    public LeavesWorld(UUID worldId, Map<ChunkPosition, LeavesChunk> chunks) {
        this.worldId = worldId;
        this.chunks = chunks;
    }

    public @UnmodifiableView Map<ChunkPosition, LeavesChunk> getChunks() {
        return Collections.unmodifiableMap(this.chunks);
    }

    public @Nullable LeavesChunk getChunk(ChunkPosition position) {
        return this.chunks.get(position);
    }

    public void editLeavesChunk(ChunkPosition position, Consumer<@Nullable LeavesChunk> consumer) {
        final LeavesChunk chunk = this.chunks.get(position);
        consumer.accept(chunk);
    }

    public void editInsertLeavesChunk(
            ChunkPosition position,
            Consumer<LeavesChunk> consumer,
            Function<ChunkPosition, LeavesChunk> function
    ) {
        final LeavesChunk chunk = this.chunks.computeIfAbsent(position, function);
        consumer.accept(chunk);
    }

    public UUID worldId() {
        return this.worldId;
    }

}

File: common/src/main/kotlin/com/hibiscusmc/hmcleaves/world/LeavesWorld.kt
package com.hibiscusmc.hmcleaves.world

import com.hibiscusmc.hmcleaves.config.LeavesConfig
import com.hibiscusmc.hmcleaves.database.LeavesDatabase
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class LeavesWorld(
    val database: LeavesDatabase,
    val world: UUID,
    private val chunks: MutableMap<ChunkPosition, LeavesChunk> = ConcurrentHashMap()
) {

    fun reload(leavesConfig: LeavesConfig) {
        this.chunks.values.forEach { chunk -> chunk.reload(leavesConfig) }
    }

    operator fun get(position: ChunkPosition): LeavesChunk? {
        return this.chunks[position]
    }

    fun getOrAdd(
        position: ChunkPosition,
        supplier: (key: ChunkPosition) -> LeavesChunk = { LeavesChunk(this.database, it) }
    ): LeavesChunk {
        return this.chunks.computeIfAbsent(position, supplier)
    }

    fun remove(position: ChunkPosition): LeavesChunk? {
        return this.chunks.remove(position)
    }

}